tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_packing.cc

```diff
@@ -0,0 +1,321 @@
+         ++i) {
+      memoryDistribution.insert(std::make_pair(i, alignedBufferSegment));
+      userangeIDs.emplace_back(i);
+    }
+
+    allocToUserangeIDs.insert(std::make_pair(alloc, userangeIDs));
+    valueOffset.insert(std::make_pair(
+        alloc, std::make_pair((size_t)0, numAlignedSegments - 1)));
+  }
+
+  /// Compute the 64 byte alinged segments of a given Value.
+  size_t computeAlignedSegments(Value v) {
+    size_t padding = 64;
+    size_t bytes = computeByteSize(v);
+    return std::ceil(bytes / (double)padding);
+  }
+
+  /// Compute the byte size of a given Value.
+  size_t computeByteSize(Value v) {
+    return v.getType().cast<ShapedType>().getSizeInBits() / 8;
+  }
+
+  /// Compute the potential reusers from the given PackingInfo vector. If the
+  /// useIntervals interfere and the aligned buffer size of the other
+  /// PackingInfo is bigger, we continue. Otherwise we order the potential
+  /// reusers by their aligned buffer size.
+  void computePotentialReuser(std::vector<PackingInfo> &other) {
+    for (auto info : other) {
+      Value v = info.alloc;
+      // Skip if the other PackingInfo is this PackingInfo.
+      if (alloc == v)
+        continue;
+
+      // If the useIntervals of this PackingInfo and the other PackingInfo
+      // interfere, we cannot reuse the other PackingInfo.
+      if (userange.getLastUseIndex(alloc) >= info.totalUseInterval.first &&
+          userange.getFirstUseIndex(alloc) <= info.totalUseInterval.second)
+        continue;
+
+      if (info.numAlignedSegments > numAlignedSegments)
+        continue;
+
+      // Find the correct place to insert the potential reuser based on the
+      // aligned buffer size.
+      auto iter = potentialReuser.begin();
+      while (iter != potentialReuser.end() &&
+             iter->numAlignedSegments >= info.numAlignedSegments)
+        ++iter;
+
+      potentialReuser.insert(iter, info);
+    }
+  }
+
+  /// Try to merge all potential reuser in the current buffer and compute a new
+  /// buffer memory distribution.
+  void packPotentialReuser(DenseSet<Value> &alreadyUnited) {
+    // Iterate over all potential reusers and check if possible candidates fit
+    // into buffer.
+    for (PackingInfo &candidate : potentialReuser) {
+      // Create a copy of the current memory distribution.
+      DenseMap<size_t, std::vector<PackingOffsetPair>> possibleMemeoryDist(
+          memoryDistribution);
+      DenseMap<Value, std::pair<size_t, size_t>> possibleOffsetToAdd;
+      size_t possibleLastUsedByte = candidate.lastUsedByte;
+
+      // Check if the candidate is already packed into an other buffer.
+      if (alreadyUnited.contains(candidate.alloc))
+        continue;
+      // Flag if we can unite the buffers.
+      bool canUnite = true;
+
+      // Iterate over allocs to check if we can allocate the same buffer
+      // segments for all userangeIDs.
+      for (auto &cInterval : candidate.allocToUserangeIDs) {
```

Comment:
Would it be better to express this as a list of intervals instead of expanding it into a vector of start positions? For large allocated sizes, the number of start positions will be big and expressing them as (start,end) intervals saves space. Also, in essence this is the same logic as checking interference on useranges with gaps, maybe some of that can be reused? Both are sequences of intervals.